### ISSUE TITLE
with exactly same env, some host cannot launch horovod proc with parallelism

### ISSUE THREAD

**Issue by inscite** [uber/horovod#168]
Greetings!
Recently, I installed horovod on 4 different nodes with same environment factors.
However, only one node returns segmentation fault error.

OS: CentOS 7.1
Python: Anaconda Python 3.6.2 w/ virtualenv
TF: TensorFlow r1.4 source build for CUDA 9.0
CUDA: CUDA 9.0 w/ cuDNN v7
GPUs: multiple P100, K40m
O2IB: Mellanox Infiniband

Here's my command list for build OpenMPI:
$ sudo -s
$ ./configure --prefix=/usr/local --with-openib=/usr/src/ofa_kernel
$ make -j 40 install
$ source activate <virtualenv>
$ pip install horovod-*.tar.gz

Installation was successful, and I launched test MNIST code which was provided at this repository. (tensorflow_mnist.py)
I successfully parallelized with 2 P100s/4 K40ms with below code:
mpirun -np 6 -H host1:2,host2:2,host3:2 -bind-to none -map-by slot -x NCCL_DEBUG=DEBUG -x LD_LIBRARY_PATH bash -c 'source activate <virtualenv>;readlink -f ./;python tensorflow_mnist.py'

But when I append a problematic node on this syndicate, the process just terminated right after execution.
So, I tested single node validation on the problematic node.
Single process with single process? Works like a charm with command below:
mpirun -np 1 -H hater:1 -bind-to none -map-by slot -x NCCL_DEBUG=DEBUG -x LD_LIBRARY_PATH bash -c 'source activate <virtualenv>;python tensorflow_mnist.py'

Unfortunately, if I extend this workbench with two processes (2 GPUs), it crashes!
Command: mpirun -np 2 -H hater:2 -bind-to none -map-by slot -x NCCL_DEBUG=DEBUG -x LD_LIBRARY_PATH bash -c 'source activate <virtualenv>;python tensorflow_mnist.py'

Here's the full print of error message:
Extracting /<dreamland>/horovod-mnist-dataset/train-images-idx3-ubyte.gz
Extracting /<dreamland>/horovod-mnist-dataset/train-images-idx3-ubyte.gz
Extracting /<dreamland>/horovod-mnist-dataset/train-labels-idx1-ubyte.gz
Extracting /<dreamland>/horovod-mnist-dataset/t10k-images-idx3-ubyte.gz
Extracting /<dreamland>/horovod-mnist-dataset/t10k-labels-idx1-ubyte.gz
Extracting /<dreamland>/horovod-mnist-dataset/train-labels-idx1-ubyte.gz
Extracting /<dreamland>/horovod-mnist-dataset/t10k-images-idx3-ubyte.gz
Extracting /<dreamland>/horovod-mnist-dataset/t10k-labels-idx1-ubyte.gz
WARNING:tensorflow:From tensorflow_mnist.py:85: get_or_create_global_step (from tensorflow.contrib.framework.python.ops.variables) is deprecated and will be removed in a future version.
Instructions for updating:
Please switch to tf.train.get_or_create_global_step
WARNING:tensorflow:From tensorflow_mnist.py:85: get_or_create_global_step (from tensorflow.contrib.framework.python.ops.variables) is deprecated and will be removed in a future version.
Instructions for updating:
Please switch to tf.train.get_or_create_global_step
[hater:58418] *** Process received signal ***
[hater:58418] Signal: Segmentation fault (11)
[hater:58418] Signal code: Invalid permissions (2)
[hater:58418] Failing at address: 0x7f54c0e7d240
[hater:58418] [ 0] /usr/lib64/libpthread.so.0(+0xf130)[0x7f54c0aa6130]
[hater:58418] [ 1] [0x7f54c0e7d240]
[hater:58418] *** End of error message ***
bash: line 1: 58418 Segmentation fault      (core dumped) python tensorflow_mnist.py
-------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
-------------------------------------------------------
--------------------------------------------------------------------------
mpirun detected that one or more processes exited with non-zero status, thus causing
the job to be terminated. The first process to do so was:

  Process name: [[63012,1],0]
  Exit code:    139
--------------------------------------------------------------------------

Here's my question: What can I do right now? (such as openmpi build with cuda flags...)
Thx!

**Comment by alsrgv** [comment]
@inscite, can you do `gdb python /path/to/core/dump`, then `thread apply all bt` and share the stack traces?

**Comment by alsrgv** [comment]
@inscite, looks like the stack trace is pointing to InfiniBand plugin for Open MPI.  Can you try adding `-mca btl ^openib` to your `mpirun` command, which would disable IB at MPI level for a test?  It would still keep using IB for data exchange in NCCL.

**Comment by inscite** [comment]
@alsrgv  Certainly. Here's how I generated several log files:
$ mpirun -mca btl ^openib -np 2 -H localhost:2 -bind-to none -map-by slot -x NCCL_DEBUG=DEBUG -x LD_LIBRARY_PATH bash -c 'source activate mpiworker;python tensorflow_mnist.py'

Well, the process also dead as a result. Here's some gdb stack traces:
[gdb_out_core_128354.log](https://github.com/uber/horovod/files/1710304/gdb_out_core_128354.log)
[gdb_out_core_128359.log](https://github.com/uber/horovod/files/1710305/gdb_out_core_128359.log)

Thx!

**Comment by alsrgv** [comment]
Very strange that it crashes on one particular machine.  Couple of things to check:
1. Are there any environment differences? OS, kernel, python version, etc.
2. Are there any messages in `dmesg` / syslog?

**Comment by inscite** [comment]
@alsrgv For both you and me, this is a bizarre situation!

There was no significant message on dmesg, as well as any system logs.
Both Kernel and OS are exactly same, as well as python version is same.

Well, due to the unexpected issue, we expected to reset that exact computing node soon.
If there exists any issue again or not, we are happy to share the information about that node.
Thank you for helping us about this case.

Sincerely,
inscite

**Comment by inscite** [comment]
After cleaning the problematic computing node, horovod works seamlessly.
I can't figure out why the node was not available at the previous moment.
Anyway, I am closing this issue! Thx ;-)
